### PR TITLE
[SPARK-27280][SQL]infer more filter from join or condition

### PR DESCRIPTION
**Pull request by windpiger**
## What changes were proposed in this pull request?
current spark can handle this query to infer fitler from join condition:
```
select * from t1 join t2 where ((t1.a = t2.d and t1.b = 1) or (t1.a = t2.d and t1.b = 5))

(t1.b =1 or t1.b =5) can be infered, and pushed to table t1.
```
but it can not handle following query:
```
select * from t1 join t2 where (t1.a = t2.d and t1.b = 1 and t2.e=3 ) or (t1.a = t2.d and t1.b = 5 and t2.e=7)

(t1.b=1 or t1.b=5) should be infered and push to t1
(t2.e=3 or t2.e=7) should be infered and push to t2

```

## How was this patch tested?
unit test added

### CODE REVIEW

**Comment by beliefer** [comment]
Is this SQL grammar supported by Hive or standard SQL?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -903,6 +910,50 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan]
       Filter(newPredicates.reduce(And), plan)
     }
   }
+
+  /**
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Is this SQL grammar supported by Hive or standard SQL?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: hive can do this

**Comment by AngersZhuuuu** [comment]
#24973  I want to do the same thing ...

**Comment by maropu** [comment]
I also think we need not a band-aid fix, but a general solution for this issue as @cloud-fan suggested in #24973. Have you checked the previous activity for the CNF normalization? https://issues.apache.org/jira/browse/SPARK-6624

**Comment by AngersZhuuuu** [comment]
> I also think we need not a band-aid fix, but a general solution for this issue as @cloud-fan suggested in #24973. Have you checked the previous activity for the CNF normalization? https://issues.apache.org/jira/browse/SPARK-6624

thanks for Issue link. I will work on this in partition pureing part.

**Pull request closed without merging**